feat(ChromaDB Node): Add local chromadb support for complete local vector database

KaanAydinli opened the pull request:
## Summary

Add ChromaDB vector store node to address vector store gaps in local RAG development. This integration provides essential vector database capabilities for local development workflows. 

The node supports standard vector store operations including load, insert, retrieve, update, and retrieve-as-tool modes with collection management and metadata filtering capabilities. Includes proper error handling for embedding dimension mismatches and collection management features.

Future work includes cloud ChromaDB support and integration with default embedding functions for improved developer experience.

<img width="913" height="590" alt="Screenshot from 2025-09-21 01-09-58" src="https://github.com/user-attachments/assets/514adc41-5b6c-4d50-a9f2-1afe23bafbf0" />


## Review / Merge checklist
- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Remove an extraneous trailing line from `pnpm-workspace.yaml`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b2621744ee81ea9c1c4ac752b8632947ca44450e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

KaanAydinli commented on packages/@n8n/nodes-langchain/nodes/vector_store/VectorStoreChromaDB/VectorStoreChromaDB.node.ts:
```diff
@@ -0,0 +1,190 @@
+		};
+
+		try {
+			const config: ChromaLibArgs = { collectionName: collection };
```
It does not ignore, it adds to config if user provides chromaURL as can be seen from code block. 
			if (options.chromaURL) {
				config.clientParams = {
					host: new URL(options.chromaURL).hostname,
					port: new URL(options.chromaURL).port
						? parseInt(new URL(options.chromaURL).port, 10)
						: 8000,
				};
			}

Joffcom commented:
Hey @KaanAydinli,

Thank you for your contribution. We appreciate the time and effort you’ve taken to submit this pull request.

Before we can proceed, please ensure the following:
	•	Tests are included for any new functionality, logic changes or bug fixes.
	•	The PR aligns with our contribution guidelines.

Regarding new nodes:
We no longer accept new nodes directly into the core codebase. Instead, we encourage contributors to follow our Community Node Submission Guide to publish nodes independently.

If your node integrates with an AI service that you own or represent, please email nodes@n8n.io and we will be happy to discuss the best approach.

About review timelines:
This PR has been added to our internal tracker as "GHC-4532". While we plan to review it, we are currently unable to provide an exact timeframe. Our goal is to begin reviews within a month, but this may change depending on team priorities. We will reach out when the review begins.

Thank you again for contributing to n8n.

felipebrunosousa10-wq commented on packages/@n8n/nodes-langchain/nodes/vector_store/VectorStoreChromaDB/VectorStoreChromaDB.node.ts:
```diff
@@ -0,0 +1,199 @@
+					console.error('ChromaDB collection listing error:', error);
+					return { results: [] };
+				}
+			},
```
import { ChromaClient } from '@chroma-core/client';

export class VectorStoreChromaDB {
  constructor(private chromaUrl: string) {}

  async insertDocuments(documents: any[]) {
    // Inicializa o cliente Chroma com o URL configurado pelo usuário
    const chromaClient = new ChromaClient({
      url: this.chromaUrl || 'http://localhost:8000', // fallback para default
    });

    // Inserir documentos no servidor correto
    await chromaClient.insertDocuments(documents);
  }
}

Joffcom commented:
Hey @KaanAydinli 

I have closed this PR as it doesn't follow our contribution guidelines, in this case it is the submission of a new node which we are currently not accepting unless it is something we have talked about previously with you.

michael-radency commented:
Hello @KaanAydinli , PR does not have credentials, so I'm receiving ChromaConnectionError:

```
(Use `node --trace-deprecation ...` to show where the warning was created)
ChromaDB collection listing error: ChromaConnectionError: Failed to connect to chromadb. Make sure your server is running and try again. If you are running from a browser, make sure that your chromadb instance is configured to allow requests from the current origin using the CHROMA_SERVER_CORS_ALLOW_ORIGINS environment variable.
```

also branch has merge conflicts

KaanAydinli commented:
@michael-radency hello again, I updated it to include credentials for local and cloud as well. But Aikido Security always fails and I couldnt find the reason for it. I sanitize the inputs for credentials. Can you help me identify the issue? Also if there are other problems you can mention them as well and I will start working on them.

KaanAydinli commented:
@michael-radency I fixed the Aikido Security problem and credentials. Also removed unnecessary libraries causing this licence issue. I’d appreciate your PR review whenever you have time. Also do you want me to add tests for the node?

KaanAydinli commented:
@michael-radency Should i add suppression for console.warn()'s from ChromaDB. It usually complains about the embed functions and really makes the console messy. n8n does warn when there isnt any embedding model connected to the vector store so Chromadb does not need to worry about not having an embedding function. I mean both works fine for me but I just wanted to inform.

michael-radency reviewed (changes requested):
Thanks for updates @KaanAydinli , I did left some comments, regarding embeddings you did mention that it is coming in the future but it does looks like we need to include support for this in this PR

KaanAydinli commented on packages/@n8n/nodes-langchain/credentials/ChromaApi.credentials.ts:
```diff
@@ -0,0 +1,160 @@
+				},
+			],
+			default: 'selfHosted',
+		},
```
I split them across 2 credentials for self-hosted and cloud API. Also used the n8n's pattern for selecting from these 2 credentials

KaanAydinli commented on packages/@n8n/nodes-langchain/credentials/ChromaApi.credentials.ts:
```diff
@@ -0,0 +1,160 @@
+					name: 'Cloud',
+					value: 'cloud',
+					description: 'Connect to Chroma Cloud',
+				},
```
It was caused by Langchain trying to use default ChromaClient to connect to cloud. I added a check for apikey and if it is present then it uses CloudClient instead of default ChromaClient. 


```typescript
if ('apiKey' in clientParams && clientParams.apiKey) {
    // Use CloudClient for Chroma Cloud
    this.index = new CloudClient({
        apiKey: clientParams.apiKey as string,
        tenant: clientParams.tenant as string | undefined,
        database: clientParams.database as string | undefined,
    });
} else {
    // Use ChromaClient for self-hosted instances
    const { ChromaClient } = await ExtendedChroma.imports();
    const clientConfig = this.url ? { path: this.url, ...clientParams } : clientParams;

    this.index = new ChromaClient(clientConfig);
}
```

KaanAydinli commented:
How can they fail i reverted the exact commit which they all used to pass

KaanAydinli committed:
fix(ChromaDB): Split credentials into Self-Hosted and Cloud types and fix Cloud connection

KaanAydinli committed:
fix(Chroma Node): Update packages

michael-radency committed:
Merge branch 'master' of https://github.com/n8n-io/n8n into pr/KaanAydinli/19806

michael-radency committed:
lock fix

michael-radency committed:
fix lock file

michael-radency committed:
lint fixes

michael-radency commented:
Hello @KaanAydinli,

I’ve pushed a commit that fixes the lock file and addresses some lint issues. Please let me know when the PR is ready for review. Before requesting review, make sure there are no lint warnings, all tests are passing, and the operations work correctly in both self-hosted and cloud modes.

If possible, a short video demonstrating that the node and credentials are working would help speed up the review process. Since this node will be included in n8n, please add comments with documentation for both the node and the credentials. You can use the documentation of existing nodes as a reference.

If you run into merge conflicts in the future, I recommend using the following commands:

```
git fetch origin
git checkout origin/master -- pnpm-lock.yaml
```


If the n8n remote is not called `origin`, adjust the command accordingly. Then run:

`pnpm install && pnpm build --force`

michael-radency reviewed (changes requested):
please add tests for the node

KaanAydinli committed:
Fix lint issues  and local connection list errors

KaanAydinli committed:
Edit description of base url in credentials

KaanAydinli committed:
Suppress Chroma warnings as we handle them

KaanAydinli committed:
Add tests for Chroma node

KaanAydinli commented:
Here is the video me testing the node in both cloud and local database with insert and retrieve modes.

[n8nChroma.webm](https://github.com/user-attachments/assets/f287af75-f2dd-4e83-93f1-3bdde645ce0c)

KaanAydinli committed:
Edit the modes

philipithomas commented on packages/@n8n/nodes-langchain/credentials/ChromaCloudApi.credentials.ts:
```diff
@@ -0,0 +1,57 @@
+
+	test: ICredentialTestRequest = {
+		request: {
+			baseURL: 'https://api.trychroma.com',
```
I recommend setting this as the default URL, but allowing an override. We will support additional data centers in the future.

KaanAydinli committed:
Add baseurl  to properties

KaanAydinli committed:
Change bearer token to x-chroma-token and add tests

KaanAydinli committed:
Fix lint and format issues

michael-radency committed:
Merge branch 'master' of https://github.com/n8n-io/n8n into pr/KaanAydinli/19806

michael-radency reviewed (dismissed):
Thanks for the amazing work, @KaanAydinli! This will be a great addition to n8n.

michael-radency committed:
cleanup

michael-radency committed:
cleanup

michael-radency committed:
Merge branch 'master' into chroma

michael-radency committed:
Merge branch 'master' into chroma

michael-radency committed:
Merge branch 'master' into chroma
